project_task_analytic_propagation/models/account_analytic_line.py

```diff
@@ -1,24 +1,45 @@
-# Copyright 2024 Moduon Team S.L.
+# Copyright 2025 Moduon Team S.L.
 # License LGPL-3.0 or later (https://www.gnu.org/licenses/LGPL-3.0)
+from contextlib import suppress
+
 from odoo import api, fields, models
 
 
 class AccountAnalyticLine(models.Model):
     _inherit = "account.analytic.line"
 
     account_id = fields.Many2one(
-        compute="_compute_account_id",
-        store=True,
+        compute="_compute_analytic_accounts",
         readonly=False,
+        store=True,
     )
 
-    @api.depends("task_id.analytic_account_id")
-    def _compute_account_id(self):
-        """Change the analytic account non billed lines."""
-        timesheet_ids = self.filtered(
-            lambda t: not t.is_so_line_edited
-            and t._is_not_billed()
-            and t.task_id.analytic_account_id
-        )
-        for timesheet in timesheet_ids:
-            timesheet.account_id = timesheet.task_id.analytic_account_id
+    def _get_analytic_plan_columns(self):
+        return [
+            f"task_id.{x}"
+            for x in self.task_id._get_plan_fnames()
+            if self._fields.get(x)
+        ]
+
+    @api.depends(lambda self: self._get_analytic_plan_columns())
+    def _compute_analytic_accounts(self):
+        plan_fnames = self._get_plan_fnames()
+        for line in self:
+            if line.is_so_line_edited:
+                continue
+            with suppress(KeyError):
+                for fname in plan_fnames:
+                    if line.task_id[fname]:
+                        line[fname] = line.task_id[fname]
+                    elif line.task_id.project_id[fname]:
+                        line[fname] = line.task_id.project_id[fname]
+                    else:
+                        line[fname] = False
```

Comment:
suggestion: with that last change, this can be very reduced:


```suggestion
                    line[fname] = line.task_id[fname] or.task_id.project_id[fname]
```

Reply:
The issue with that suggestion is that we want to set unset some plans for some tasks timesheets

Reply:
That case is covered AFAICS. If `line.task_id[fname] or.task_id.project_id[fname]` returns `False` or an empty recordset, it will still be set in the field.